### ISSUE TITLE
how to use tensorrt with python

### ISSUE THREAD

**Issue by hailcpy** [wang-xinyu/tensorrtx#155]
I followed your instuction for arcface and converted to tensorrt. It is working great with 6 time less inference time. Can you help me to use this with python. The onnx parser doesn't work for me as i have tensorrt 7 which needs onnx >=1.6.0 and the mxnet arcface model require onnx <=1.4.1. I can't change my tensorrt version. Your code worked fine. So if you could help me in running this with python. Thanks

**Comment by wang-xinyu** [comment]
You can go through the getting started tutorial of this repo and try to write the build engine part yourself.

**Comment by dongdongqiang2018** [comment]
Hi, how to use engine of Retinaface to inference with python. @wang-xinyu

**Comment by wang-xinyu** [comment]
@dongdongqiang2018 also refer to the yolov5 python wrapper.

**Comment by dongdongqiang2018** [comment]
I imported engine with reference to yolov5_trt.py, but it failed
![微信图片_20200826111636](https://user-images.githubusercontent.com/49979601/91251260-a0a82300-e78d-11ea-8e6b-ecd63a5e9ddd.png)

**Comment by wang-xinyu** [comment]
@dongdongqiang2018 please try the latest commit, I just fix it. Clean build/ dir and rebuild everything.

**Comment by dongdongqiang2018** [comment]
Still doesn't seem to work

**Comment by wang-xinyu** [comment]
what' s the error now?

**Comment by dongdongqiang2018** [comment]
like this
![微信图片_20200826111636](https://user-images.githubusercontent.com/49979601/91268783-7dd73800-e7a8-11ea-8c91-5a5d408303ab.png)

**Comment by wang-xinyu** [comment]
did you rebuild the libdeocdeplugin.so?

**Comment by dongdongqiang2018** [comment]
yes, I pulled this item again, just use the previous wts file

**Comment by wang-xinyu** [comment]
I already move the `REGISTER_TENSORRT_PLUGIN(DecodePluginCreator);` to the decode.h,   it won't report error like `could not find plugin Decode_TRT version 1` if rebuild the libdecodeplugin.so properly.

no idea...

**Comment by dongdongqiang2018** [comment]
Ok i will try again

**Comment by hailcpy** [comment]
@wang-xinyu  what preprocessing did you do before passing the image to arcface. I am taking the output of mtcnn model from insightface and directly feeding it to arcface which is giving weird similarity ratings. The output of mtcnn is (1,3,112,112). Should i change something. Also i have written the python file for arcface replicating that from yolov5. After this problem is solved i will add it to your repo. Thanks.

**Comment by wang-xinyu** [comment]
```
    cv::Mat img = cv::imread("../joey0.ppm");
    for (int i = 0; i < INPUT_H * INPUT_W; i++) {
        data[i] = ((float)img.at<cv::Vec3b>(i)[2] - 127.5) * 0.0078125;
        data[i + INPUT_H * INPUT_W] = ((float)img.at<cv::Vec3b>(i)[1] - 127.5) * 0.0078125;
        data[i + 2 * INPUT_H * INPUT_W] = ((float)img.at<cv::Vec3b>(i)[0] - 127.5) * 0.0078125;
    }
```

The arcface here only did normalization in preprocessing.

In your application, you need use something like mtcnn to detect facial landmarks and do alignment to get 112x112 face.

**Comment by hailcpy** [comment]
hey @wang-xinyu  I am taking the output of get_input() from insightface/deploy/face_model.py and then feeding it to arcface. The output from get_input() contains the aligned image and the size of image is 112x112.

**Comment by dongdongqiang2018** [comment]
After switching from the server environment to the local environment,  recompile it again, and now the error becomes this
![微信图片_20200826111636](https://user-images.githubusercontent.com/49979601/91276419-1c1ccb00-e7b4-11ea-958b-07002db6a807.png)
Will it possible for you to share your engine and .so files ?

**Comment by hailcpy** [comment]
Hey i solved the issue. Will add pull request for tensorrt python api. Thanks for your contribution and help.

**Comment by dongdongqiang2018** [comment]
Will it possible for you to share your .so verison?@wang-xinyu

**Comment by wang-xinyu** [comment]
what's the error now?

**Comment by dongdongqiang2018** [comment]
I did it in this way
Resnet50_Final.pth  ->  onnx -> trt_engine
then inference...

I want to compare the difference between the two methods

**Comment by jiagh2010** [comment]
Resnet50_Final.pth -> onnx -> trt_engine   @dongdongqiang2018   Is this a success？ Is the result consistent with the official website？

**Comment by dongdongqiang2018** [comment]
form  https://github.com/biubug6/Pytorch_Retinaface download pretrain model.
pth2onnx like this
```
#torch version == 1.6.0
import os
os.environ["CUDA_VISIBLE_DEVICES"] = "1"
from models.retinaface import RetinaFace
from data import cfg_mnet, cfg_re50
import torch

def check_keys(model, pretrained_state_dict):
    ckpt_keys = set(pretrained_state_dict.keys())
    model_keys = set(model.state_dict().keys())
    used_pretrained_keys = model_keys & ckpt_keys
    unused_pretrained_keys = ckpt_keys - model_keys
    missing_keys = model_keys - ckpt_keys
    print('Missing keys:{}'.format(len(missing_keys)))
    print('Unused checkpoint keys:{}'.format(len(unused_pretrained_keys)))
    print('Used keys:{}'.format(len(used_pretrained_keys)))
    assert len(used_pretrained_keys) > 0, 'load NONE from pretrained checkpoint'
    return True


def remove_prefix(state_dict, prefix):
    ''' Old style model is stored with all names of parameters sharing common prefix 'module.' '''
    print('remove prefix \'{}\''.format(prefix))
    f = lambda x: x.split(prefix, 1)[-1] if x.startswith(prefix) else x
    return {f(key): value for key, value in state_dict.items()}


def load_model(model, pretrained_path, load_to_cpu):
    print('Loading pretrained model from {}'.format(pretrained_path))
    if load_to_cpu:
        pretrained_dict = torch.load(pretrained_path, map_location=lambda storage, loc: storage)
    else:
        device = torch.cuda.current_device()
        pretrained_dict = torch.load(pretrained_path, map_location=lambda storage, loc: storage.cuda(device))
    if "state_dict" in pretrained_dict.keys():
        pretrained_dict = remove_prefix(pretrained_dict['state_dict'], 'module.')
    else:
        pretrained_dict = remove_prefix(pretrained_dict, 'module.')
    check_keys(model, pretrained_dict)
    model.load_state_dict(pretrained_dict, strict=False)
    return model


model = 'weights/Resnet50_Final.pth'
cpu = False

net = RetinaFace(cfg_re50, phase = 'test')
net = load_model(net, model, cpu)
net.eval()
print('Finished loading model!')
print(net)
device = torch.device("cpu" if cpu else "cuda")
net = net.to(device)

output_onnx = 'FaceDetector.onnx'
print("==> Exporting model to ONNX format at '{}'".format(output_onnx))
input_names = ["input0"]
output_names = ["output0"]
inputs = torch.randn(1, 3, 736,1280).to(device)

torch_out = torch.onnx.export(net, inputs, output_onnx, export_params=True, opset_version=11,verbose=True,
                                input_names=input_names, output_names=output_names)
```
next,  get engine

```
onnx2trt FaceDetector.onnx -o resnet50.engine
```

The two results are the same

**Comment by jiagh2010** [comment]
@dongdongqiang2018   LResNet50E-IR,ArcFace@ms1m-refine-v1 ** Consistent with its results?

**Comment by dongdongqiang2018** [comment]
sorry,I only used resnet50

**Comment by Vuong02011996** [comment]
@hailcpy Have you create a pull request for ArcFace Python API ? Could you please share with me the code ? I'll really appreciate it. Thank you. 
Best regards,
Vuong

**Comment by hailcpy** [comment]
hey @Vuong02011996  i did create an inference script but there is one problem because of which i didn't do the pull request. I am sharing the link of the script. The script works fine if you display the results as you do inference on images one by one, but if you try to store the embeddings in some list the last embedding vector overwrites all the existing embedding vectors in the list. I don't know what is the reason for this problem. I did some changes and used it for face recognition with tensorrt mtcnn detector from jk_jung and since i am not storing and embedding vectors, it's not an issue for me.
Here https://drive.google.com/file/d/1v_1WdNad2_olQbOS6Q4TqJg60dyuDfK_/view?usp=sharing

**Comment by thimabru1010** [comment]
> hey @Vuong02011996 i did create an inference script but there is one problem because of which i didn't do the pull request. I am sharing the link of the script. The script works fine if you display the results as you do inference on images one by one, but if you try to store the embeddings in some list the last embedding vector overwrites all the existing embedding vectors in the list. I don't know what is the reason for this problem. I did some changes and used it for face recognition with tensorrt mtcnn detector from jk_jung and since i am not storing and embedding vectors, it's not an issue for me.
> Here https://drive.google.com/file/d/1v_1WdNad2_olQbOS6Q4TqJg60dyuDfK_/view?usp=sharing

How did you managed to extract landmarks and bboxes fromthis script. After reshaping like output.reshape(-1, 1) I observe a (1, 189001) shape vector. How does I obtain theses infos from your script?

**Comment by weixiu00** [comment]
> > hey @Vuong02011996 i did create an inference script but there is one problem because of which i didn't do the pull request. I am sharing the link of the script. The script works fine if you display the results as you do inference on images one by one, but if you try to store the embeddings in some list the last embedding vector overwrites all the existing embedding vectors in the list. I don't know what is the reason for this problem. I did some changes and used it for face recognition with tensorrt mtcnn detector from jk_jung and since i am not storing and embedding vectors, it's not an issue for me.
> > Here https://drive.google.com/file/d/1v_1WdNad2_olQbOS6Q4TqJg60dyuDfK_/view?usp=sharing
> 
> How did you manage to extract landmarks and bboxes from this script? After reshaping like output.reshape(-1, 1) I observe a (1, 189001) shape vector. How do I obtain these infos from your script?

Hi @thimabru1010 i had the same situation. did you figure out how to extract landmarks and bboxes from the (1, 189001) shape vector ?
